### PR TITLE
Make weather station list interactive

### PR DESCRIPTION
**Pull request by floatplane**
The user-facing part of this PR is pretty basic - navigating from the weather station list to a (placeholder) station detail page:

https://user-images.githubusercontent.com/101196/215909237-4e6f8318-dbbf-46a3-bf91-f501c998e561.mp4

The underlying React Navigation bits were soooooo painful though. I'll add comments inline.

### CODE REVIEW

File: components/forecast/WeatherTab.tsx [hunk 1]
```diff
@@ -5,14 +5,22 @@ import {AllCapsSm, AllCapsSmBlack, Body, BodyBlack, BodySemibold, bodySize, Titl
...
+import {apiDateString, utcDateToLocalTimeString} from 'utils/date';
 import {InfoTooltip} from 'components/content/InfoTooltip';
 import helpStrings from 'content/helpStrings';
 import {useWeatherStations} from 'hooks/useWeatherStations';
+import {ActionList} from 'components/content/ActionList';
+import {CompositeNavigationProp, useNavigation} from '@react-navigation/native';
+import {HomeStackParamList, TabNavigationProps} from 'routes';
+import {NativeStackNavigationProp} from '@react-navigation/native-stack';
+
+type ForecastNavigationProp = CompositeNavigationProp<NativeStackNavigationProp<HomeStackParamList, 'forecast'>, TabNavigationProps>;
```
[the remark this answers is not in the document]
Comment (floatplane) [reply components/forecast/WeatherTab.tsx:18]: I guess I should say "this is _part of_ the black magic" - the changes in `routes.ts` are also required 🪄

File: App.tsx [hunk 2]
```diff
@@ -167,17 +167,11 @@ const BaseApp: React.FunctionComponent<{
...
-                {state => TelemetryTabScreen(withParams(state, {center_id: avalancheCenterId, dateString: dateString}))}
-              </TabNavigator.Screen>
-              <TabNavigator.Screen name="Menu" initialParams={{center_id: avalancheCenterId}}>
-                {() => MenuStackScreen(avalancheCenterId, setAvalancheCenterId, staging, setStaging)}
+              <TabNavigator.Screen name="Home">{() => <HomeTabScreen defaultCenterId={avalancheCenterId} defaultDateString={dateString} />}</TabNavigator.Screen>
+              <TabNavigator.Screen name="Observations">{() => <ObservationsTabScreen defaultCenterId={avalancheCenterId} defaultDateString={dateString} />}</TabNavigator.Screen>
+              <TabNavigator.Screen name="Weather Data">{() => <WeatherScreen defaultCenterId={avalancheCenterId} defaultDateString={dateString} />}</TabNavigator.Screen>
+              <TabNavigator.Screen name="Menu">
+                {() => <MenuStackScreen avalancheCenterId={avalancheCenterId} setAvalancheCenter={setAvalancheCenterId} staging={staging} setStaging={setStaging} />}
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply App.tsx:174]: Do we lose the ability to deep-link to the tabs, though? What part of this was required by the rest of the changes?

File: components/content/ActionList.tsx [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  action?: (label: string, data: T) => void;
+  // Someday, might be nice to support nesting
+  // children?: Action[];
+}
+
+export interface ActionListProps<T> extends ViewProps {
+  actions: Action<T>[];
+}
+
+export function ActionList<T>({actions, ...props}: ActionListProps<T>) {
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply components/content/ActionList.tsx:20]: Do we want to add an icon or something to make it more obvious that these are links?

File: components/forecast/AvalancheForecast.tsx [hunk 4]
```diff
@@ -126,7 +126,7 @@ export const AvalancheForecast: React.FunctionComponent<AvalancheForecastProps>
           <AvalancheTab zone={zone} forecast={forecast} />
         </Tab>
         <Tab title="Weather">
-          <WeatherTab zone={zone} />
+          <WeatherTab zone={zone} center_id={center_id} date={date} />
```
Comment (stevekuznetsov) [inline components/forecast/AvalancheForecast.tsx:129]: Still kind of weird that we're not passing a zone here, the screen is very specifically for the zone we're looking at the avy forecast for, right?

File: components/forecast/WeatherTab.tsx [hunk 5]
```diff
@@ -144,16 +153,23 @@ export const WeatherTab: React.FC<WeatherTabProps> = ({zone}) => {
...
+          {weatherStationStatus === 'success' && groupedWeatherStations.length > 0 && (
+            <ActionList
+              actions={groupedWeatherStations.map(([name, stations]) => ({
+                label: name,
+                data: stations,
+                action: () => {
+                  // Nested navigation to the stationDetail page of the Weather Data stack
+                  navigation.navigate('Weather Data', {
+                    screen: 'stationDetail',
+                    // Treat this as the first screen in the Weather Data stack - don't show a back button going to the stationList
```
Comment (stevekuznetsov) [inline components/forecast/WeatherTab.tsx:165]: How does the user get back to the weather tab?

File: hooks/useWeatherStations.ts [hunk 6]
```diff
@@ -68,10 +68,6 @@ const decommissionedStations = [
 ];
 
 export const useWeatherStations = ({center, sources}: Props) => {
-  if (center !== 'NWAC') {
-    throw new Error(`can't fetch weather for ${center}: useWeatherStations hook only supports NWAC`);
-  }
-
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply hooks/useWeatherStations.ts:74]: It's ok :)

File: routes.ts [hunk 7]
```diff
@@ -56,3 +63,11 @@ export type MenuStackParamList = {
   textStylePreview: undefined;
 };
 export type MenuStackNavigationProps = NativeStackNavigationProp<MenuStackParamList>;
+
+export type TabNavigatorParamList = {
+  Home: NavigatorScreenParams<HomeStackParamList>;
+  'Weather Data': NavigatorScreenParams<WeatherStackParamList>;
+  Observations: NavigatorScreenParams<ObservationsStackParamList>;
+  Menu: NavigatorScreenParams<MenuStackParamList>;
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply routes.ts:71]: Yeah the hoisting definitely occurs - so this approach necessarily means that you _couldn't_ navigate to this tab by itself - could a composite be used here, though, to retain that? Would we ever expect to deep-link to a tab?

File: components/content/ActionList.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (floatplane) [reply components/content/ActionList.tsx:20]: That's a design question, but to me the ">" is a really common pattern for drilling into a detail view.

File: routes.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (floatplane) [reply routes.ts:71]: > could a composite be used here, though, to retain that?

I don't know, tbh I really ran out of gas at the end of the day yesterday and didn't want to battle the compiler anymore! But it really _seems_ like we should be able to support that, in which case a lot of the App.tsx changes could get reverted. I'll poke at it (in a timeboxed way).

> Would we ever expect to deep-link to a tab?

It feels somewhat theoretical 🤷. Each of these tabs has a stack navigator, and each stack navigator has its own default screen (set through `initialRouteName`), so in some sense you're never just displaying the tab - you're _always_ displaying the tab with a particular subpath. Maybe a question is, do we need to?

File: components/forecast/AvalancheForecast.tsx (hunk 4 above)
Comment (floatplane) [reply components/forecast/AvalancheForecast.tsx:129]: Isn't `zone` the very first param there? 😕

File: components/forecast/WeatherTab.tsx (hunk 5 above)
Comment (floatplane) [reply components/forecast/WeatherTab.tsx:165]: We talked about this a bit in the design meeting, there isn't a design yet. Currently you have to tap the home icon in the tab bar (so you're not stuck).

File: App.tsx [hunk 8]
```diff
@@ -167,14 +169,14 @@ const BaseApp: React.FunctionComponent<{
                   }
                 },
               })}>
-              <TabNavigator.Screen name="Home" initialParams={{center_id: avalancheCenterId, dateString: dateString}}>
-                {state => HomeTabScreen(withParams(state, {center_id: avalancheCenterId, dateString: dateString}))}
+              <TabNavigator.Screen name="Home" initialParams={{center_id: avalancheCenterId, dateString}}>
+                {state => HomeTabScreen(merge(state, {route: {params: {center_id: avalancheCenterId, dateString}}}))}
```
Comment (floatplane) [inline App.tsx:173]: The compiler was convinced that with `withParams`, we were losing the `screen` field of `params` (which comes in via `NavigationScreenParams`). `merge` is designed for this job anyway and seems to not offend tsc.

File: routes.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply routes.ts]: I'd love to put deep links to the "home tab" or whatever on the website - but if we can instead just link to the default route in the stack, I don't think it matters "how" it's done.

File: components/forecast/AvalancheForecast.tsx (hunk 4 above)
Comment (stevekuznetsov) [reply components/forecast/AvalancheForecast.tsx:129]: :facepalm:

File: components/forecast/WeatherTab.tsx [hunk 9]
```diff
@@ -144,16 +153,26 @@ export const WeatherTab: React.FC<WeatherTabProps> = ({zone}) => {
...
+            <ActionList
+              actions={groupedWeatherStations.map(([name, stations]) => ({
+                label: name,
+                data: stations,
+                action: () => {
+                  // Nested navigation to the stationDetail page of the Weather Data stack
+                  const dateString = apiDateString(date);
+                  navigation.navigate('Weather Data', {
+                    center_id,
+                    dateString,
```
Comment (floatplane) [inline components/forecast/WeatherTab.tsx:166]: slight change from previous revision - now `center_id` and `dateString` have to be passed as params for both the `Weather Data` screen, and the `stationDetail` screen. kinda weird, but livable I think.

File: components/forecast/WeatherTab.tsx (hunk 5 above)
Comment (stevekuznetsov) [reply components/forecast/WeatherTab.tsx:168]: Oh sorry, I had understood it to mean there were back arrows to let you get back to the forecast, e.g. weather station -> weather tab / forecast tab view

**Review by floatplane — COMMENTED** [review]
ptal @stevekuznetsov 

this is a much simpler diff now!

**Comment by stevekuznetsov** [comment]
LGTM! Happy to have lodash do the typescript nonsense for us :)

**Pull request merged**